src/main/java/com/deviceinsight/kafka/health/KafkaConsumingHealthIndicator.java

```diff
@@ -135,64 +159,65 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
 		}
 	}
 
-	private KafkaCommunicationResult sendAndReceiveMessage() {
+	private void sendMessage() {
 
 		Future<Void> sendReceiveTask = null;
 
 		try {
 
 			sendReceiveTask = executor.submit(() -> {
-				sendAndReceiveKafkaMessage();
+				sendKafkaMessage();
 				return null;
 			});
 
 			sendReceiveTask.get(sendReceiveTimeoutMs, MILLISECONDS);
+			this.kafkaCommunicationResult = KafkaCommunicationResult.success(topic);
 
 		} catch (ExecutionException e) {
 			logger.warn("Kafka health check execution failed.", e);
-			return KafkaCommunicationResult.failure(topic, e);
+			this.kafkaCommunicationResult = KafkaCommunicationResult.failure(topic, e);
 		} catch (TimeoutException | InterruptedException e) {
 			logger.warn("Kafka health check timed out.", e);
 			sendReceiveTask.cancel(true);
-			return KafkaCommunicationResult.failure(topic, e);
+			this.kafkaCommunicationResult = KafkaCommunicationResult.failure(topic, e);
 		} catch (RejectedExecutionException e) {
 			logger.debug("Ignore health check, already running...");
 		}
-		return KafkaCommunicationResult.success(topic);
 	}
 
-	private void sendAndReceiveKafkaMessage() throws Exception {
+	private void sendKafkaMessage() throws Exception {
 
 		String message = UUID.randomUUID().toString();
 		long startTime = System.currentTimeMillis();
 
 		logger.debug("Send health check message = {}", message);
-		producer.send(new ProducerRecord<>(topic, message, message)).get(sendReceiveTimeoutMs, MILLISECONDS);
 
-		while (messageNotReceived(message)) {
-			logger.debug("Waiting for message={}", message);
-		}
+		producer.send(new ProducerRecord<>(topic, message, message)).get(sendReceiveTimeoutMs, MILLISECONDS);
+		cacheService.write(message);
 
 		logger.debug("Kafka health check succeeded. took= {} msec", System.currentTimeMillis() - startTime);
 	}
 
-
-	private boolean messageNotReceived(String message) {
+	private boolean messageNotReceived() {
 
 		return StreamSupport.stream(consumer.poll(Duration.ofMillis(pollTimeoutMs)).spliterator(), false)
-				.noneMatch(msg -> msg.key().equals(message) && msg.value().equals(message));
+				.noneMatch(msg -> cacheService.get(msg.key()) == null);
 	}
 
 
 	@Override
 	protected void doHealthCheck(Health.Builder builder) {
-		KafkaCommunicationResult kafkaCommunicationResult = sendAndReceiveMessage();
+		sendMessage();
```

Comment:
Wouldn't it be enough to send the Kafka message and use the last `KafkaCommunicationResult`?